allennlp/semparse/action_space_walker.py

```diff
@@ -0,0 +1,117 @@
+from collections import defaultdict
+from typing import List, Dict, Set
+
+from allennlp.semparse.worlds.world import World
+from allennlp.semparse.type_declarations import type_declaration as types
+
+
+class ActionSpaceWalker:
+    """
+    ``ActionSpaceWalker`` takes a world, traverses all the valid paths driven by the valid action
+    specification of the world to generate all possible logical forms (under some constraints). This
+    class also has some utilities for indexing logical forms to efficiently retrieve required
+    subsets.
+
+    Parameters
+    ----------
+    world : ``World``
+        The world from which valid actions will be taken.
+    max_path_length : ``int``
+        The maximum path length till which the action space will be explored. Paths longer than this
+        length will be discarded.
+    """
+    def __init__(self, world: World, max_path_length: int) -> None:
+        self._world = world
+        self._max_path_length = max_path_length
+        self._completed_paths: List[List[str]] = None
+        self._terminal_path_index: Dict[str, Set[int]] = defaultdict(set)
+        self._length_sorted_paths: List[List[str]] = None
+
+    def _walk(self) -> None:
+        """
+        Walk over action space to collect completed paths of at most ``self._max_path_length`` steps.
+        """
+        # Buffer of NTs to expand, previous actions
+        incomplete_paths = [([str(type_)], [f"@START@ -> {type_}"]) for type_ in
```

Comment:
I know that this is mainly a utility class but can we use the `START_SYMBOL` rather than a string, particularly as here it only has single surrounding `@`s

Reply:
Yep, you're right.  Done.